File: LeetCodeSolutions/LeetCode_0046.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return nums
        n = len(nums)
        res = []
        tmp = [None for i in nums]
        used = [0 for i in nums]

        def permute_all(cur_idx):
            if cur_idx == n:
                res.append(tmp[:])
                return
            for i in range(n):
                if not used[i]:
                    tmp[cur_idx] = nums[i]
                    used[i] = 1
                    permute_all(cur_idx + 1)
                    used[i] = 0

        permute_all(0)
        return res

File: LeetCodeSolutions/LeetCode_0046.cpp
class Solution {
public:
    void backtrack(int idx, vector<int> &nums, vector<vector<int>> &res, vector<int> &tmp, vector<bool> &used){
        if (idx == used.size()){
            res.push_back(tmp);
            return ;
        }
        for (int i=0; i!=used.size(); ++i){
            if (!used[i]){
                used[i] = true;
                tmp.push_back(nums[i]);
                backtrack(idx + 1, nums, res, tmp, used);
                used[i] = false;
                tmp.pop_back();
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.empty()) return res;
        int sz = nums.size();
        vector<int> tmp = {};
        vector<bool> used(sz, false);
        backtrack(0, nums, res, tmp, used);
        return res;
    }
};